### ISSUE TITLE
Strange convergence behavior if using CG

### ISSUE THREAD

**Issue by hr87** [idaholab/moose#8681]
### Description of the enhancement or error report
I have a SPD system and want to take advantage by using the CG solver. However, the solve is noticeable slower than with GMRES and the convergence shows a linear residual of 0, but continues to iterate. Exodiff confirms that both GMRES and CG give the same result, however CG takes many times the number of iterations.

### Rationale for the enhancement or information for reproducing the error

Modifying the example `test/tests/kernels/2d_diffusion/2d_diffusion_test.i` to

```
[Mesh]
  file = square.e
[]

[Variables]
  active = 'u'

  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  active = 'diff'

  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  active = 'left right'

  [./left]
    type = DirichletBC
    variable = u
    boundary = 1
    value = 0
  [../]

  [./right]
    type = DirichletBC
    variable = u
    boundary = 2
    value = 1
  [../]
[]

[Executioner]
  type = Steady

  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type'
  petsc_options_value = 'hypre boomeramg cg'

[]

[Outputs]
  file_base = out
  exodus = true
[]
```

CG convergence:
```
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 0.000000e+00
      1 Linear |R| = 0.000000e+00
      2 Linear |R| = 0.000000e+00
      3 Linear |R| = 0.000000e+00
      4 Linear |R| = 0.000000e+00
      5 Linear |R| = 0.000000e+00
      6 Linear |R| = 0.000000e+00
 1 Nonlinear |R| = 1.769225e-08
      0 Linear |R| = 0.000000e+00
      1 Linear |R| = 0.000000e+00
      2 Linear |R| = 0.000000e+00
      3 Linear |R| = 0.000000e+00
      4 Linear |R| = 0.000000e+00
      5 Linear |R| = 0.000000e+00
      6 Linear |R| = 0.000000e+00
      7 Linear |R| = 0.000000e+00
      8 Linear |R| = 0.000000e+00
      9 Linear |R| = 0.000000e+00
     10 Linear |R| = 0.000000e+00
 2 Nonlinear |R| = 0.000000e+00
```
GMRES convergence:
```
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 1.732051e+00
      1 Linear |R| = 4.836823e-02
      2 Linear |R| = 3.331048e-16
 1 Nonlinear |R| = 1.055803e-08
```

**Comment by YaqiWang** [comment]
Thanks Hans. I will take a quick look tomorrow. Possibly @fdkong is interested in this.

**Comment by YaqiWang** [comment]
Looks like MOOSE is using the right preconditioning by default which prevent PETSc from evaluating the residual norm. Running with the following can fix this issue:
```
../../../moose_test-opt -i 2d_diffusion_test.i -ksp_type cg Preconditioning/smp/type=SMP Preconditioning/smp/pc_side=left
```
but this is definitely undesirable.

**Comment by fdkong** [comment]
CG should work with both left and right preconditioners. I guess the matrix may not be  symmetric due to boundary conditions even-though the operator is self adjoint. Let me take a quick check.

**Comment by fdkong** [comment]
```
row 0: (0, 1.)  (1, 0.)  (2, 0.)  (3, 0.) 
row 1: (0, -0.166667)  (1, 1.33333)  (2, -0.333333)  (3, -0.333333)  (4, -0.166667)  (5, -0.333333) 
row 2: (0, -0.333333)  (1, -0.333333)  (2, 2.66667)  (3, -0.333333)  (4, -0.333333)  (5, -0.333333)  (6, -0.333333)  (7, -0.333333)  (8, -0.333333) 
row 3: (0, 0.)  (1, 0.)  (2, 0.)  (3, 1.)  (6, 0.)  (7, 0.) 
row 4: (1, 0.)  (2, 0.)  (4, 1.)  (5, 0.) 
row 5: (1, 0.)  (2, 0.)  (4, 0.)  (5, 1.)  (6, 0.)  (8, 0.) 
row 6: (2, -0.333333)  (3, -0.333333)  (5, -0.333333)  (6, 1.33333)  (7, -0.166667)  (8, -0.166667) 
row 7: (2, 0.)  (3, 0.)  (6, 0.)  (7, 1.) 
row 8: (2, 0.)  (5, 0.)  (6, 0.)  (8, 1.) 
```

You can see that this matrix is not  symmetric. For example, row 1: (5, -0.333333) is different from row 5: (1, 0).  

If you really want to use CG, you have to move the Dirichlet boundary conditions to the right hand side and reduce the dimension of the matrix (to get a symmetric matrix).  MOOSE may not support this right now. @jwpeterson may know more on this.

**Comment by jwpeterson** [comment]
Yes, we enforce Dirichlet BCs non-symmetrically in MOOSE, there is currently no other option. 

The non-symmetric approach is more efficient when running large parallel jobs since after assembly of the preconditioner or Jacobian matrix, you only modify one row per constrained boundary DOF rather than every row... 

Supposing you could get a symmetric matrix, what kind of time savings do you expect to see in using CG through PETSc vs. GMRES?  Isn't the effectiveness of the preconditioner the most important thing in the end?

**Comment by YaqiWang** [comment]
Dirichlet BC can be replaced with VacuumBC to maintain symmetry. Looks like this is not a convergence issue, just residual norm is not evaluated and shown.

**Comment by fdkong** [comment]
Barry told me that PETSc CG is not supported with right preconditioning, although it mathematically makes sense to do it in the right.  We may support this in the future. 

If you want to use CG, please use the left preconditioning with "-ksp_pc_side right".

**Comment by YaqiWang** [comment]
That option is currently not working because of https://github.com/idaholab/moose/blob/devel/framework/src/utils/PetscSupport.C#L468.
This is the reason why I had to add a SMP on the command line and change the PC side there. If we can honor command line `-ksp_pc_side left`  or in `petsc_options` in the input file, it will be ok.

**Comment by fdkong** [comment]
> This is the reason why I had to add a SMP on the command line and change the PC side there. If we can honor command line -ksp_pc_side left or in petsc_options in the input file, it will be ok.

Both` -ksp_pc_side left `or ` petsc_options` are working. Not working on your machine?

**Comment by YaqiWang** [comment]
You are right, it is working! I must skewed up something.

**Comment by hr87** [comment]
@jwpeterson The main reason to use CG is the memory issue. CG requires the storage of only 3 solution vectors compared to 20 or even 100 for GMRES depending on the restart value. For neutronics calculations with over 1000 variables one can imagine that this makes a large impact of needed RAM. Hence preserving the SPD matrix and actually using CG allow to run larger problems. The convergence of CG should be at least the same as GMRES.

**Comment by jwpeterson** [comment]
> The main reason to use CG is the memory issue. 

Ah, I did not think of the memory side of things. 

It would probably require some effort to allow MOOSE to do symmetric Dirichlet BC enforcement on an element-by-element basis, although it is currently possible to do this in libmesh, see for example `introduction_ex4`.  The main idea is to use a `DirichletBoundary` object and call 
```
dof_map.heterogenously_constrain_element_matrix_and_vector (Ke, Fe, dof_indices);
```
when assembling the Jacobian/preconditioner matrices.  It is probably better to try something like @YaqiWang suggested instead to keep the problem symmetric.

**Comment by YaqiWang** [comment]
Should we then add few lines in `petscSetDefaults()` to change the preconditioning side to left for certain KSP types?

Maybe we should have something like:
```
switch (ksp_type)
{
  case GMRES:
    ...
  case CG:
   ...
  default:
   mooseError("unsupported KSP type");
}
```

**Comment by fdkong** [comment]
> Should we then add few lines in petscSetDefaults() to change the preconditioning side to left for certain KSP types?
> 
> Maybe we should have something like:
> 
> switch (ksp_type)
> {
>   case GMRES:
>     ...
>   case CG:
>    ...
>   default:
>    mooseError("unsupported KSP type");
> }

NO. This already has been done in PETSc. I do not know why we have KSPSetPCSide in MOOSE. @andrsd  may know this. It is better to use default settings in PETSc.

**Comment by YaqiWang** [comment]
We need to call this because we have a parameter `Preconditioning/*/pc_side` and we need to bring it in. The problem is that when that parameter is not set, we are using right preconditioning by default with
https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystemBase.C#L123

**Comment by fdkong** [comment]
> We need to call this because we have a parameter Preconditioning/*/pc_side and we need to bring it in. The problem is that when that parameter is not set, we are using right preconditioning by default with
> https://github.com/idaholab/moose/blob/devel/framework/src/base/NonlinearSystemBase.C#L123

I think we should do NOTHING when ` Preconditioning/*/pc_side` is NOT set, instead of using right preconditioning, and then default  values in PETSc will be used. It is better to let PETSc make the decision.

Any reason to set the right side as the default value in MOOSE?

**Comment by permcody** [comment]
Tagging @friedmud

**Comment by fdkong** [comment]
@permcody, I want to make a change. If  `Preconditioning/*/pc_side ` is NOT set, I want to let PETSc make the decision. The reason is that different solvers in PETSc come with different default settings. The right preconditioning is invalid  for some solvers in PETSc, e.g., CG.

**Comment by andrsd** [comment]
The reason for right preconditioning as a default was that it was the thing we are doing by default back in the day.

Feel free to fix the code so that if nothing is specified, then PETSc will make the decision and if something is set than that value will be used.

**Comment by andrsd** [comment]
@hr87 You can try to use `PenaltyDirichletBC` which should enforce the Dirichlet-type BC weakly and should preserve the symmetricity.

**Comment by jwpeterson** [comment]
There is also [Nitsche's method for enforcing Dirichlet BCs](http://www.ams.org/journals/mcom/2009-78-267/S0025-5718-08-02183-2/) which I was just reading about recently.  It maintains the symmetry of the matrix and is better-conditioned than the penalty method. Not sure if it applies to your problem.

**Comment by YaqiWang** [comment]
`PenaltyDirichletBC` is the one I was in mind (I called it `VacuumBC`, but realized `VacuumBC` does not have nonhomogeneous term in it). Nitsche's method is more like a nature extension of interior penalty method by allowing the difference between u and u_0. We are adopting a method more like the `PenaltyDirichletBC`. But anyway, we do not have the strong imposed Dirichlet BC, so non-symmetric Dirichlet BCs in MOOSE is not a problem to us. Thanks @jwpeterson and @andrsd .

**Comment by fdkong** [comment]
`PenaltyDirichletBC` has a bug, I think.

```
row 0: (0, 16667.3)  (1, -0.166667)  (2, -0.333333)  (3, 8333.17) 
row 1: (0, -0.166667)  (1, 1.33333)  (2, -0.333333)  (3, -0.333333)  (4, -0.166667)  (5, -0.333333) 
row 2: (0, -0.333333)  (1, -0.333333)  (2, 2.66667)  (3, -0.333333)  (4, -0.333333)  (5, -0.333333)  (6, -0.333333)  (7, -0.333333)  (8, -0.333333) 
row 3: (0, 8333.17)  (1, -0.333333)  (2, -0.333333)  (3, 33334.7)  (6, -0.333333)  (7, 8333.17) 
row 4: (1, -0.166667)  (2, -0.333333)  (4, 16667.3)  (5, 8333.17) 
row 5: (1, -0.333333)  (2, -0.333333)  (4, 8333.17)  (5, 33334.7)  (6, -0.333333)  (8, 8333.17) 
row 6: (2, -0.333333)  (3, -0.333333)  (5, -0.333333)  (6, 1.33333)  (7, -0.166667)  (8, -0.166667) 
row 7: (2, -0.333333)  (3, 8333.17)  (6, -0.166667)  (7, 16667.3) 
row 8: (2, -0.333333)  (5, 8333.17)  (6, -0.166667)  (8, 16667.3) 
```

We should scale the DIAG only. But here off-diag values such as `row 0: (3, 8333.17) ` are also scaled. Anyone know why? @YaqiWang  @andrsd

**Comment by andrsd** [comment]
Are you running @hr87 's input file with `DirichletBC` changed to `PenaltyDirichletBC`?

**Comment by fdkong** [comment]
> Are you running @hr87 's input file with DirichletBC changed to PenaltyDirichletBC?

Yes. Is it the wrong way to use `PenaltyDirichletBC`?

**Comment by andrsd** [comment]
Ok, I will look at it.

**Comment by jwpeterson** [comment]
> On Mar 9, 2017, at 12:37 PM, David Andrs <notifications@github.com> wrote:
> 
> Ok, I will look at it.
> 

Penalty BC typically adds a symmetric mass matrix contribution for L2 projection of the data.  So it has off-diagonal entries...

**Comment by andrsd** [comment]
So, the residual is `_p * _test[_i][_qp] * (-_v + _u[_qp]);`. The linear combination of `_u[_qp]` is using values from "neighboring" shape function and that generates the off-diagonal scaled term (because it gets multiplied by `_p`). This is with GUASS quadrature. Switching to FIRST, TRAP and only diagonal entries are scaled. 

@fdkong What do you think should be done instead?

**Comment by fdkong** [comment]
@andrsd 

How about: ?

```
Real
PenaltyDirichletBC::computeQpJacobian()
{
  if (_i == _j)
    return  _p*_phi[_j][_qp]*_test[_i][_qp];
  else
    return  0.0;  
}
```

Only the diag values are scaled.

```
row 0: (0, 16667.3)  (1, -0.166667)  (2, -0.333333)  (3, -0.166667) 
row 1: (0, -0.166667)  (1, 1.33333)  (2, -0.333333)  (3, -0.333333)  (4, -0.166667)  (5, -0.333333) 
row 2: (0, -0.333333)  (1, -0.333333)  (2, 2.66667)  (3, -0.333333)  (4, -0.333333)  (5, -0.333333)  (6, -0.333333)  (7, -0.333333)  (8, -0.333333) 
row 3: (0, -0.166667)  (1, -0.333333)  (2, -0.333333)  (3, 33334.7)  (6, -0.333333)  (7, -0.166667) 
row 4: (1, -0.166667)  (2, -0.333333)  (4, 16667.3)  (5, -0.166667) 
row 5: (1, -0.333333)  (2, -0.333333)  (4, -0.166667)  (5, 33334.7)  (6, -0.333333)  (8, -0.166667) 
row 6: (2, -0.333333)  (3, -0.333333)  (5, -0.333333)  (6, 1.33333)  (7, -0.166667)  (8, -0.166667) 
row 7: (2, -0.333333)  (3, -0.166667)  (6, -0.166667)  (7, 16667.3) 
row 8: (2, -0.333333)  (5, -0.166667)  (6, -0.166667)  (8, 16667.3) 
```

**Comment by andrsd** [comment]
Then, your jacobian is inconsistent with your residuals...

**Comment by fdkong** [comment]
> Then, your jacobian is inconsistent with your residuals...

Agreed. Let us stay with the current implementation.

**Comment by YaqiWang** [comment]
Do we have the BC in the paper given by @jwpeterson somewhere in MOOSE? If not, we should add it.

**Comment by jwpeterson** [comment]
@YaqiWang not that I know of; I was thinking about adding it just to see how it works.  It work well as a companion to the `Diffusion` Kernel already in MOOSE, but it would not be generally applicable to other second-order PDEs unless there is a way for the user to specify what the flux term is for their application.

**Comment by andrsd** [comment]
> unless there is a way for the user to specify what the flux term is for their application.

We can do an abstract class inheriting from `IntegratedBC` and add the API for people to specify the flux like we similarly do in `FluxBC` already... Or maybe inherit from `FluxBC` and reimplement `computeQp{Residual|Jacobian}`.

**Comment by permcody** [comment]
@fdkong, others. I believe this issue can be closed. Please take a look.

**Comment by lindsayad** [comment]
I've retitled this issue and it's still relevant today. @roystgnr has plans to bring the libMesh Dirichlet boundary condition constraint enforcement into MOOSE from libMesh. Additionally, and this is outside the scope of this particular issue, I would like to do something similar for (non)linear systems to what I'm doing in https://github.com/libMesh/libmesh/pull/3821 and https://github.com/idaholab/moose/pull/27276 for eigen systems which is to condense out constraint dofs leaving only "physics" dofs in the matrix we hand to PETSc

**Comment by lindsayad** [comment]
@roystgnr maybe we can reference this issue when we bring the libMesh Dirichlet boundary condition enforcement into MOOSE
